lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -4722,6 +4722,59 @@ FIRRTLType RefSubOp::inferReturnType(ValueRange operands,
       loc, "ref.sub op requires a RefType of vector or bundle base type");
 }
 
+//===----------------------------------------------------------------------===//
+// Optional Group Operations
+//===----------------------------------------------------------------------===//
+
+LogicalResult GroupOp::verify() {
+  auto groupName = getGroupName();
+  auto *parentOp = (*this)->getParentOp();
+
+  auto nestedReferences = groupName.getNestedReferences();
+  if (nestedReferences.empty()) {
+    if (!isa<FModuleOp>(parentOp)) {
+      auto diag = mlir::emitError(getLoc())
+                  << "'" << getOperationName()
+                  << "' op has an un-nested group symbol, but does not have a "
+                     "'firrtl.module' op as a parent";
+      return diag.attachNote(parentOp->getLoc())
+             << "illegal parent op defined here";
+    }
+  } else {
+    auto parentGroup = dyn_cast<GroupOp>(parentOp);
+    if (!parentGroup) {
+      auto diag = mlir::emitError(getLoc())
+                  << "'" << getOperationName()
+                  << "' op has a nested group symbol, but does not have a '"
+                  << getOperationName() << "' op as a parent'";
+      return diag.attachNote(parentOp->getLoc())
```

Comment:
Feels like these branches can be combined to share most of this text/reasoning but maybe not worth it, just a thought.